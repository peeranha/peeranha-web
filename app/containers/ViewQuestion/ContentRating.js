import React from 'react';
import PropTypes from 'prop-types';

import fingerUpSingleQuestionPage from 'images/fingerUpSingleQuestionPage.svg';
import greenFingerUpSingleQuestion from 'images/greenFingerUpSingleQuestion.svg';
import fingerDownSingleQuestionPage from 'images/fingerDownSingleQuestionPage.svg';
import redFingerDownSingleQuestion from 'images/redFingerDownSingleQuestion.svg';
import disabledFingerUp from 'images/disabledFingerUp.svg';
import disabledFingerDown from 'images/disabledFingerDown.svg';
import emptyFingerUp from 'images/emptyFingerUp.svg';
import emptyFingerDown from 'images/emptyFingerDown.svg';

import { BORDER_SUCCESS, BORDER_WARNING_LIGHT } from 'style-constants';
import { getFormattedNum } from 'utils/numbers';

import Base from 'components/Base';
import Span from 'components/Span';
import MediumImage from 'components/Img/MediumImage';

import { UP_VOTE_BUTTON, DOWN_VOTE_BUTTON } from './constants';

const MediumImageStyled = MediumImage.extend`
  ${(props /* istanbul ignore next */) =>
    props.src === greenFingerUpSingleQuestion
      ? `border: 1px solid ${BORDER_SUCCESS};`
      : ``};

  ${(props /* istanbul ignore next */) =>
    props.src === redFingerDownSingleQuestion
      ? `border: 1px solid ${BORDER_WARNING_LIGHT};`
      : ``};
`;

const ContentRating = /* istanbul ignore next */ ({
  answerId,
  account,
  upVote,
  votingStatus,
  rating,
  downVote,
  userInfo,
}) => (
  <Base className="d-flex align-items-center justify-content-between">
    <button
      className="p-0"
      onClick={upVote}
      id={`${UP_VOTE_BUTTON}${answerId}`}
      data-answerid={answerId}
      data-whowasupvoted={userInfo.user}
    >
      <UpvoteIcon
        account={account}
        userInfo={userInfo}
        votingStatus={votingStatus}
      />
    </button>

    <Span fontSize="20" bold>
      {getFormattedNum(rating)}
    </Span>

    <button
      className="p-0"
      onClick={downVote}
      id={`${DOWN_VOTE_BUTTON}${answerId}`}
      data-answerid={answerId}
      data-whowasdownvoted={userInfo.user}
    >
      <DownvoteIcon
        account={account}
        userInfo={userInfo}
        votingStatus={votingStatus}
      />
    </button>
  </Base>
);

ContentRating.propTypes = {
  rating: PropTypes.number,
  votingStatus: PropTypes.object,
  userInfo: PropTypes.object,
  account: PropTypes.string,
  upVote: PropTypes.func,
  downVote: PropTypes.func,
  answerId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
};

function UpvoteIcon({
  account,
  userInfo,
  votingStatus,
}) /* istanbul ignore next */ {
  let src = null;

  if (account === userInfo.user) {
    src = disabledFingerUp;
  } else if (votingStatus.isUpVoted) {
    src = greenFingerUpSingleQuestion;
  } else if (votingStatus.isDownVoted) {
    src = emptyFingerUp;
  } else {
    src = fingerUpSingleQuestionPage;
  }

  return <MediumImageStyled src={src} alt="voteup" />;
}

UpvoteIcon.propTypes = {
  votingStatus: PropTypes.object,
  userInfo: PropTypes.object,
  account: PropTypes.string,
};

function DownvoteIcon({
  account,
  userInfo,
  votingStatus,
}) /* istanbul ignore next */ {
  let src = null;

  if (account === userInfo.user) {
    src = disabledFingerDown;
  } else if (votingStatus.isDownVoted) {
    src = redFingerDownSingleQuestion;
  } else if (votingStatus.isUpVoted) {
    src = emptyFingerDown;
  } else {
    src = fingerDownSingleQuestionPage;
  }

  return <MediumImageStyled src={src} alt="votedown" />;
}

DownvoteIcon.propTypes = {
  votingStatus: PropTypes.object,
  userInfo: PropTypes.object,
  account: PropTypes.string,
};

export { UpvoteIcon, DownvoteIcon, ContentRating };
export default React.memo(ContentRating);
