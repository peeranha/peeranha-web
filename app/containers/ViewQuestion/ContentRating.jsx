/* eslint indent: 0 */
import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

import fingerUpSingleQuestionPage from 'images/fingerUpSingleQuestionPage.svg?inline';
import greenFingerUpSingleQuestion from 'images/greenFingerUpSingleQuestion.svg?inline';
import fingerDownSingleQuestionPage from 'images/fingerDownSingleQuestionPage.svg?inline';
import redFingerDownSingleQuestion from 'images/redFingerDownSingleQuestion.svg?inline';
import disabledFingerUp from 'images/disabledFingerUp.svg?inline';
import disabledFingerDown from 'images/disabledFingerDown.svg?inline';
import emptyFingerUp from 'images/emptyFingerUp.svg?inline';
import emptyFingerDown from 'images/emptyFingerDown.svg?inline';

import { BORDER_SUCCESS, BORDER_WARNING_LIGHT } from 'style-constants';
import { getFormattedNum } from 'utils/numbers';

import Span from 'components/Span';
import Button from 'components/Button/Contained/Transparent';

import { UP_VOTE_BUTTON, DOWN_VOTE_BUTTON } from './constants';

const ImgBox = styled.div`
  position: relative;
  display: flex;
  align-items: center;
  justify-content: center;

  :after {
    content: '';
    position: absolute;
    top: calc(50% - 21px);
    left: calc(50% - 21px);
    width: 42px;
    height: 42px;
    border-radius: 50%;

    ${x =>
      x.src === greenFingerUpSingleQuestion
        ? `border: 1px solid ${BORDER_SUCCESS};`
        : ``};

    ${x =>
      x.src === redFingerDownSingleQuestion
        ? `border: 1px solid ${BORDER_WARNING_LIGHT};`
        : ``};
  }

  @media only screen and (max-width: 576px) {
    ${x =>
      x.src === greenFingerUpSingleQuestion ||
      x.src === redFingerDownSingleQuestion
        ? `width: 42px; height: 42px;`
        : ``};
  }
`;

const ContentRating = ({
  answerId,
  account,
  upVote,
  votingStatus,
  rating,
  downVote,
  userInfo,
  ids,
}) => (
  <React.Fragment>
    <Button
      className="overflow-initial"
      onClick={upVote}
      disabled={ids.includes(`${UP_VOTE_BUTTON}${answerId}`)}
      id={`${UP_VOTE_BUTTON}${answerId}`}
      data-answerid={answerId}
      data-whowasupvoted={userInfo.user}
    >
      <UpvoteIcon
        account={account}
        userInfo={userInfo}
        votingStatus={votingStatus}
      />
    </Button>

    <Span fontSize="20" bold>
      {getFormattedNum(rating)}
    </Span>

    <Button
      className="overflow-initial"
      onClick={downVote}
      disabled={ids.includes(`${DOWN_VOTE_BUTTON}${answerId}`)}
      id={`${DOWN_VOTE_BUTTON}${answerId}`}
      data-answerid={answerId}
      data-whowasdownvoted={userInfo.user}
    >
      <DownvoteIcon
        account={account}
        userInfo={userInfo}
        votingStatus={votingStatus}
      />
    </Button>
  </React.Fragment>
);

ContentRating.propTypes = {
  rating: PropTypes.number,
  votingStatus: PropTypes.object,
  userInfo: PropTypes.object,
  account: PropTypes.string,
  upVote: PropTypes.func,
  downVote: PropTypes.func,
  ids: PropTypes.array,
  answerId: PropTypes.number,
};

function UpvoteIcon({ account, userInfo, votingStatus }) {
  let src = null;

  if (account === userInfo.user) {
    src = disabledFingerUp;
  } else if (votingStatus.isUpVoted) {
    src = greenFingerUpSingleQuestion;
  } else if (votingStatus.isDownVoted) {
    src = emptyFingerUp;
  } else {
    src = fingerUpSingleQuestionPage;
  }

  return (
    <ImgBox src={src}>
      <img src={src} alt="voteup" />
    </ImgBox>
  );
}

UpvoteIcon.propTypes = {
  votingStatus: PropTypes.object,
  userInfo: PropTypes.object,
  account: PropTypes.string,
};

function DownvoteIcon({ account, userInfo, votingStatus }) {
  let src = null;

  if (account === userInfo.user) {
    src = disabledFingerDown;
  } else if (votingStatus.isDownVoted) {
    src = redFingerDownSingleQuestion;
  } else if (votingStatus.isUpVoted) {
    src = emptyFingerDown;
  } else {
    src = fingerDownSingleQuestionPage;
  }

  return (
    <ImgBox src={src}>
      <img src={src} alt="votedown" />
    </ImgBox>
  );
}

DownvoteIcon.propTypes = {
  votingStatus: PropTypes.object,
  userInfo: PropTypes.object,
  account: PropTypes.string,
};

export { UpvoteIcon, DownvoteIcon, ContentRating };
export default React.memo(ContentRating);
