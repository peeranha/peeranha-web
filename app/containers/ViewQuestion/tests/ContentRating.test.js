import { ContentRating } from '../ContentRating';

describe('ContentRating', () => {
  const props = {
    upVote: jest.fn(),
    downVote: jest.fn(),
    votingStatus: {},
    userInfo: {
      user: 'user123',
    },
    questionData: {
      correct_answer_id: 0,
    },
  };

  it('snapshot test', () => {
    expect(ContentRating(props)).toMatchSnapshot();
  });
});
