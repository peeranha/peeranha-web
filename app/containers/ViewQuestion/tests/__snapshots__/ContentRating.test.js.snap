// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContentRating snapshot test 1`] = `
<div
  className="content-rating text-center"
>
  <MarkAsAcceptedIcon
    correct_answer_id={0}
    id="app_view_question_mark_as_buttonundefined"
    whoWasAccepted="user123"
  />
  <button
    id="app_view_question_upvote_buttonundefined"
  >
    <FontAwesomeIcon
      border={false}
      className="chevron chevron-up"
      data-voting="chevron-hl-undefined"
      fixedWidth={false}
      flip={null}
      icon="chevron-up"
      inverse={false}
      listItem={false}
      mask={null}
      onClick={[Function]}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      symbol={false}
      title=""
      transform={null}
    />
  </button>
  <span
    className="rating-value"
  >
    0
  </span>
  <button
    id="app_view_question_downvote_buttonundefined"
  >
    <FontAwesomeIcon
      border={false}
      className="chevron chevron-down"
      data-voting="chevron-hl-undefined"
      fixedWidth={false}
      flip={null}
      icon="chevron-down"
      inverse={false}
      listItem={false}
      mask={null}
      onClick={[Function]}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      symbol={false}
      title=""
      transform={null}
    />
  </button>
</div>
`;
